tracing/opentracing/http.go

```diff
@@ -0,0 +1,40 @@
+package opentracing
+
+import (
+	"net/http"
+
+	kithttp "github.com/go-kit/kit/transport/http"
+	"github.com/opentracing/opentracing-go"
+	"golang.org/x/net/context"
+)
+
+// ToHTTPRequest returns an http RequestFunc that injects an OpenTracing Span
+// found in `ctx` into the http headers. If no such Span can be found, the
+// RequestFunc is a noop.
+func ToHTTPRequest(tracer opentracing.Tracer) kithttp.RequestFunc {
+	return func(ctx context.Context, req *http.Request) context.Context {
+		// Try to find a Span in the Context.
+		if span := opentracing.SpanFromContext(ctx); span != nil {
+			// There's nothing we can do with any errors here.
+			_ = tracer.Inject(span, opentracing.TextMap, opentracing.HTTPHeaderTextMapCarrier(req.Header))
+		}
+		return ctx
+	}
+}
+
+// FromHTTPRequest returns an http RequestFunc that tries to join with an
+// OpenTracing trace found in `req` and starts a new Span called
+// `operationName` accordingly. If no trace could be found in `req`, the Span
+// will be a trace root. The Span is incorporated in the returned Context and
+// can be retrieved with opentracing.SpanFromContext(ctx).
+func FromHTTPRequest(tracer opentracing.Tracer, operationName string) kithttp.RequestFunc {
+	return func(ctx context.Context, req *http.Request) context.Context {
+		// Try to join to a trace propagated in `req`. There's nothing we can
+		// do with any errors here, so we ignore them.
+		span, _ := tracer.Join(operationName, opentracing.TextMap, opentracing.HTTPHeaderTextMapCarrier(req.Header))
+		if span == nil {
+			span = opentracing.StartSpan(operationName)
+		}
+		return opentracing.ContextWithSpan(ctx, span)
```

Comment:
hm, you're right, I just checked and for Go I didn't do much on the server side. In Python it looked like this:

```python
    span.set_tag(tags.SPAN_KIND, tags.SPAN_KIND_RPC_SERVER)
    span.set_tag('http.url', request.full_url)

    remote_ip = request.remote_ip
    if remote_ip:
        span.set_tag(tags.PEER_HOST_IPV4, remote_ip)

    caller_name = request.caller_name
    if caller_name:
        span.set_tag(tags.PEER_SERVICE, caller_name)

    remote_port = request.remote_port
    if remote_port:
        span.set_tag(tags.PEER_PORT, remote_port)
```

The `caller_name` was a custom behavior based on a special HTTP tag that our services send to identify themselves, and the instrumentation is configured with the name of that header. And Python libs usually expose remote addr to the server, similar to Java servlets. Go's `Request` also has `RemoteAddr` field, but I was lazy to parse it (note to self: don't be lazy, fix it).